### ISSUE TITLE
Flask request: RuntimeError: Working outside of request context.

### ISSUE THREAD

**Issue by mrtarunjain** [dchevell/flask-executor#1]
Environment:
I am running a flask application on gunicorn with gevent workers.
I am using flask's request variable in all the functions that I am calling through executor.
I verified that package is using app.app_context wrapper in case of Thread.
But still it is not getting flask request variable.
How to pass request variable to other threads.

Please help.

**Comment by dchevell** [comment]
Hey @mrtarunjain, do you have any example code you can share that replicates this issue?

**Comment by mrtarunjain** [comment]
Gunicorn:
```
exec gunicorn app:app \
  --name $NAME \
  --workers $NUM_WORKERS \
  -k gevent \
  --bind=0.0.0.0:80 \
  --log-level=info \
  --access-logfile=- \
  --timeout=15 \
  --reload
```

In Flask middleware I set current user id in request.
```
@app.before_request
def set_viewer_context():
    request.current_user_id = 'some-id'
```

I am calling 1 API which internally calls 4 API(that can be executed concurrently).
These 4 API's needs current user id and it gets it from request variable.

Though, I know I can do this if I start concurrent request after I have the current_user_id from request variable and then submitting this function to executor.

But is there a way where I can access current thread's request variable in the other thread?

**Comment by dchevell** [comment]
Thanks @mrtarunjain, I understand now - Flask-Executor supports wrapping your function in the _application_ context, so you can submit functions that depend on it - but not the _request_ context, which is created and destroyed inside each view function. I definitely think this library should solve for both, and I think I know exactly how we should be doing it:

http://flask.pocoo.org/docs/1.0/api/#flask.copy_current_request_context

This should give functions submitted to the executor access to both the application context _and_ the request context, so I've opened a separate branch to test this method.

I'd appreciate if you can test this yourself:

```
pip install -U git+https://github.com/dchevell/flask-executor.git@issue/1/support-request-context
```
Let me know how you go with this @mrtarunjain !

**Comment by dchevell** [comment]
@mrtarunjain I've written some tests for this that add a value to the request object, then test that same value in another thread via the executor, and it all seems to work. The results of your real world test will be great, and if it works then I'm happy to merge it and push a new version to pypi.

Thanks for the report and for helping improve this project!

**Comment by mrtarunjain** [comment]
@dchevell I tested it and it works with Flask==0.12.2. The test case you have added is same as my use case ie., reading the request variable.
What we might want to test?
- Try changing the variable in other thread and see it's effect in other threads.
- How flask variable g works?

For now, You can go ahead and merge it. It will be of great help.
Thank you so much for the changes.

**Comment by dchevell** [comment]
@mrtarunjain the way I understand it, `copy_current_request_context` doesn't give your function access to the _original_ request context, only a copy of it. As soon as it's submitted to the executor, it's a separate copy, and won't see any further changes to the original context nor propagate any local changes back to it. I'll test this to confirm of course.

I still need to write docs for this that describe the basic API and explain the expected behaviour around context handling, so I'll make sure to include this point there. For now, I'll merge this - thank you again!

**Comment by dchevell** [comment]
This is now all working and released as 0.4.0
